qiskit/transpiler/target.py

```diff
@@ -0,0 +1,652 @@
+                    out_durations.append((instruction, list(qarg), properties.duration, "s"))
+        self._instruction_durations = InstructionDurations(out_durations, dt=self.dt)
+        return self._instruction_durations
+
+    def timing_constraints(self):
+        """Get an :class:`~qiskit.transpiler.TimingConstraints` object from the target
+
+        Returns:
+            TimingConstraints: The timing constraints represented in the Target
+        """
+        return TimingConstraints(
+            self.granularity, self.min_length, self.pulse_alignment, self.aquire_alignment
+        )
+
+    def instruction_schedule_map(self):
+        """Return an :class:`~qiskit.pulse.InstructionScheduleMap` for the
+        instructions in the target with a pulse schedule defined.
+
+        Returns:
+            InstructionScheduleMap: The instruction schedule map for the
+            instructions in this target with a pulse schedule defined.
+        """
+        if self._instruction_schedule_map is not None:
+            return self._instruction_schedule_map
+        out_inst_schedule_map = InstructionScheduleMap()
+        for instruction, qargs in self._gate_map.items():
+            for qarg, properties in qargs.items():
+                if properties is not None and properties.calibration is not None:
+                    out_inst_schedule_map.add(instruction, qarg, properties.calibration)
+        self._instruction_schedule_map = out_inst_schedule_map
+        return out_inst_schedule_map
+
+    def operation_from_name(self, instruction):
+        """Get the operation class object for a given name
+
+        Args:
+            instruction (str): The instruction name to get the
+                :class:`~qiskit.circuit.Instruction` instance for
+        Returns:
+            Instruction: The Instruction instance corresponding to the name
+        """
+        return self._gate_name_map[instruction]
+
+    def operations_for_qargs(self, qarg):
+        """Get the operation class object for a specified qarg
+
+        Args:
+            qarg (tuple): A qarg tuple of the qubits to get the gates that apply
+                to it. For example, ``(0,)`` will return the set of all
+                instructions that apply to qubit 0.
+        Returns:
+            list: The list of :class:`~qiskit.circuit.Instruction` instances
+            that apply to the specified qarg.
+
+        Raises:
+            KeyError: If qarg is not in target
+        """
+        if qarg not in self._qarg_gate_map:
+            raise KeyError(f"{qarg} not in target.")
+        return [self._gate_name_map[x] for x in self._qarg_gate_map[qarg]]
+
+    @property
+    def operation_names(self):
+        """Get the operation names in the target."""
+        return set(self._gate_map)
+
+    @property
+    def operations(self):
+        """Get the operation class objects in the target."""
+        return list(self._gate_name_map.values())
```

Comment:
Perhaps `tuple` (or just raw `.values()`) is better here, because it'll suitably error if a user tries to mutate the property.  Probably it's overkill here, though, because there's nothing we can do to stop a user from screwing with the class instances contained within.